include/float.h

```diff
@@ -16,6 +27,104 @@ extern "C" {
 bool binary64_to_integer(uint64_t value, long long *n);
 bool binary64_to_unsigned(uint64_t value, unsigned long long *n);
 
+enum {
+    BINARY64_MAX_FORMAT_SPACE = 25
+};
+
+/*
+ * Convert a number in IEEE 754 binary64 floating point format to a
+ * concise, NUL-terminated decimal representation that converts back
+ * to the original value. The chosen format aims to be suitable for
+ * human use in that exponential notation is used only for "very big"
+ * or "very small" values.
+ *
+ * Special values are formatted to "-0.0", "nan", "-nan", "infinity"
+ * and "-infinity".
+ *
+ * There are no bounds checks for buffer. The binary64_format_space()
+ * function can be used to determine sufficient worst-case space
+ * requirements.
+ *
+ * The returned value is strlen(3) of the result.
+ */
+size_t binary64_format(uint64_t value, char buffer[BINARY64_MAX_FORMAT_SPACE]);
+
+/*
+ * Return a dynamically allocated, NUL-terminated string in the format
+ * given out by binary64_format() (q.v.). The return value should be
+ * freed with fsfree().
+ */
+char *binary64_to_string(uint64_t value);
+
+typedef enum {
+    BINARY64_TYPE_NAN,
+    BINARY64_TYPE_INFINITY,
+    BINARY64_TYPE_ZERO,
+    BINARY64_TYPE_NORMAL,
+} binary64_type_t;
+
+/*
+ * Convert the binary64 floating point value to decimal, exponential
+ * components. The components depend on the value type. The "negative"
+ * argument is set for all value types. The other arguments are only
+ * defined for BINARY64_TYPE_NORMAL.
+ *
+ * If case of BINARY64_TYPE_NORMAL, the components map back to value
+ * as follows:
+ *
+ *    sprintf(buffer, "%s%llue%d",
+ *            *negative ? "-" : "",
+ *            (unsigned long long) *significand,
+ *            (int) *exponent);
+ *    value == strtod(buffer, NULL);
+ *
+ * Additionally, the significand is guaranteed not to be divisible by
+ * 10.
+ */
+void binary64_to_decimal(uint64_t value,
+                         binary64_type_t *type,
+                         bool *negative,
+                         uint64_t *significand,
+                         int32_t *exponent);
```

Comment:
What about introducing a struct for the decomposed number, so that `binary64_{from,to}_decimal ` users would need to manage a single variable?

Reply:
Indeed, such a struct is used internally.